src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -438,4 +466,84 @@ protected void createHardLink(PathFragment linkPath, PathFragment originalPath)
       throws IOException {
     super.createHardLink(linkPath, originalPath);
   }
+
+  static class RemoteInMemoryFileSystem extends InMemoryFileSystem {
+
+    public RemoteInMemoryFileSystem(DigestHashFunction hashFunction) {
+      super(hashFunction);
+    }
+
+
+    @Override
+    protected FileInfo newFile(Clock clock, PathFragment path) {
+      return new RemoteFileInfo(clock);
+    }
+
+    void injectRemoteFile(PathFragment path, byte[] digest, long size, String actionId)
+        throws IOException {
+      createDirectoryAndParents(path.getParentDirectory());
+      InMemoryContentInfo node = getOrCreateWritableInode(path);
+      if (!(node instanceof RemoteFileInfo)) {
+        throw new IOException("Could not inject into " + node);
+      }
+
+      RemoteFileInfo remoteFileInfo = (RemoteFileInfo) node;
+      remoteFileInfo.set(digest, size, actionId);
+    }
+
+    @Nullable
+    RemoteFileInfo getRemoteFileInfo(PathFragment path, boolean followSymlinks) {
+      InMemoryContentInfo node = inodeStatErrno(path, followSymlinks).inode();
+      if (!(node instanceof RemoteFileInfo)) {
+        return null;
+      }
+      return (RemoteFileInfo) node;
+    }
+  }
+
+  static class RemoteFileInfo extends FileInfo {
```

Comment:
Will it be possible to reuse `InMemoryFileInfo` once we get rid of `actionId` (as we discussed elsewhere)?

Reply:
No:
1. we have to override `getFastDigest()` to return the digest of remote file.
2. we want to block operations that don't make sense to remote files e.g. `getOutputStream`.